trace/config.go

```diff
@@ -135,71 +185,56 @@ func (o attributeSpanOption) apply(c *SpanConfig) {
 // If multiple of these options are passed the attributes of each successive
 // option will extend the attributes instead of overwriting. There is no
 // guarantee of uniqueness in the resulting attributes.
-func WithAttributes(attributes ...attribute.KeyValue) LifeCycleOption {
-	return attributeSpanOption(attributes)
+//nolint:golint
+func WithAttributes(attributes ...attribute.KeyValue) attributeOption {
+	return attributeOption(attributes)
 }
 
-type timestampSpanOption time.Time
+type timestampOption time.Time
 
-func (o timestampSpanOption) ApplySpan(c *SpanConfig)  { o.apply(c) }
-func (o timestampSpanOption) ApplyEvent(c *SpanConfig) { o.apply(c) }
-func (timestampSpanOption) private()                   {}
-func (o timestampSpanOption) apply(c *SpanConfig)      { c.Timestamp = time.Time(o) }
+func (o timestampOption) applySpan(c *SpanConfig)      { c.timestamp = time.Time(o) }
+func (o timestampOption) applySpanStart(c *SpanConfig) { o.applySpan(c) }
+func (o timestampOption) applySpanEnd(c *SpanConfig)   { o.applySpan(c) }
+func (o timestampOption) applyEvent(c *EventConfig)    { c.timestamp = time.Time(o) }
 
-// WithTimestamp sets the time of a Span life-cycle moment (e.g. started,
-// stopped, errored).
-func WithTimestamp(t time.Time) LifeCycleOption {
-	return timestampSpanOption(t)
-}
-
-type linksSpanOption []Link
+var _ SpanOption = timestampOption{}
+var _ SpanEndOption = timestampOption{}
+var _ EventOption = timestampOption{}
 
-func (o linksSpanOption) ApplySpan(c *SpanConfig) { c.Links = append(c.Links, []Link(o)...) }
-func (linksSpanOption) private()                  {}
+// WithTimestamp sets the time of a Span or Event life-cycle moment (e.g.
+// started, stopped, errored).
+//nolint:golint
+func WithTimestamp(t time.Time) timestampOption {
```

Comment:
Same here, this is returning an unexported type instead of an exported option interface.